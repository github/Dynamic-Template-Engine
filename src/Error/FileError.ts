export class EmptyFileError extends Error {
  constructor(message: string) {
    super(message);
    this.name = EmptyFileError.name;
  }
}

export class FileReadError extends Error {
  constructor(message: string) {
    super(message);
    this.name = FileReadError.name;
  }
}

export class FileParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = FileParseError.name;
  }
}
