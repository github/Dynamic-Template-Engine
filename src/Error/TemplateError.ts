export class TemplateNotFound extends Error {
  constructor(message: string) {
    super(message);
    this.name = TemplateNotFound.name;
  }
}

export class TemplateParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = TemplateParseError.name;
  }
}

export class TemplateEngineNotFound extends Error {
  constructor(message: string) {
    super(message);
    this.name = TemplateEngineNotFound.name;
  }
}

export class TemplateRenderError extends Error {
  constructor(message: string) {
    super(message);
    this.name = TemplateRenderError.name;
  }
}
