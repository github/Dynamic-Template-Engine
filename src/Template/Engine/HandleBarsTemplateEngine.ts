// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/dynamic-template-engine/blob/master/LICENSE) for details)
import * as Handlebars from 'handlebars';
import { FunctionalityNotSupportedError, CustomHelperRegisterError } from '../../Error/FunctionalityError';
import ITemplateEngine from '../Core/ITemplateEngine';
import { TemplateNotFound, TemplateParseError } from '../../Error/TemplateError';

export default class HandleBarsTemplateEngine implements ITemplateEngine {
  constructor() {
    this.preCompiledTemplateMap = new Map<string, HandlebarsTemplateDelegate<any>>();
  }

  /**
   * Registers the template with the template engine by compiling and storing the compiled method.
   * TODO :: Add parials template support
   *
   * @param {string} templateId - id used to store the precompiled template
   * @param {string} template - template to register
   * @param {JSON} options - provide options such as partial or not
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public registerTemplate(templateId: string, template: string, options?: JSON): void {
    // TODO :: Add parials template support using options
    let preCompiledTemplate: HandlebarsTemplateDelegate;
    try {
      preCompiledTemplate = Handlebars.compile(template);
    } catch (error) {
      throw new TemplateParseError(error.message);
    }
    this.preCompiledTemplateMap.set(templateId, preCompiledTemplate);
  }

  /**
   * Apply the template using the data provided.
   *
   * @param {string} templateId - id with which the compiled template is stored
   * @param {JSON} dataModel - data to apply to the template
   */
  public applyTemplate(templateId: string, dataModel: JSON): string {
    const preCompiledTemplate = this.preCompiledTemplateMap.get(templateId);
    if (!preCompiledTemplate) {
      throw new TemplateNotFound(`No template found for the given templateId : ${templateId}`);
    }
    return preCompiledTemplate(dataModel);
  }

  /**
   * Register custom helper functions with template engine.
   *
   * @param helperName name of the helper to register
   * @param helperFunc the implementation of helper function
   */
  // eslint-disable-next-line class-methods-use-this
  public registerHelper(helperName: string, helperFunc: Handlebars.HelperDelegate): void {
    try {
      Handlebars.registerHelper(helperName, helperFunc);
    } catch (error) {
      throw new CustomHelperRegisterError(`Registration of custom helper: ${helperName} failed with ERROR: ${error.message} `);
    }
  }

  /**
   * Register custom tag with template engine.
   *
   * @throws FunctionalityNotSupportedError if the engine does not support custom tags/extensions
   */
  // eslint-disable-next-line class-methods-use-this
  public registerTag(): void {
    throw new FunctionalityNotSupportedError('HandleBars does not support custom tags or extensions');
  }

  private preCompiledTemplateMap: Map<string, HandlebarsTemplateDelegate<any>>;
}
