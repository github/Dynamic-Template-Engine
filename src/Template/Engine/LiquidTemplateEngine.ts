/** @copyright Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/dynamic-template-engine/blob/master/LICENSE) for details) */
import {
  Liquid, Template, FilterImplOptions, TagImplOptions,
} from 'liquidjs';
import { CustomHelperRegisterError, CustomTagRegisterError } from '../../Error/FunctionalityError';
import ITemplateEngine from '../Core/ITemplateEngine';
import { TemplateNotFound, TemplateParseError } from '../../Error/TemplateError';

export default class LiquidTemplateEngine implements ITemplateEngine {
  constructor() {
    this.preCompiledTemplateMap = new Map<string, Template[]>();
    this.engine = new Liquid();
  }

  /**
   * Registers the template with the template engine by compiling and storing the compiled method.
   * TODO :: Add parials template support using options
   *
   * @param {string} templateId - id used to store the precompiled template
   * @param {string} template - template to register
   * @param {JSON} options - provide options such as partial or not
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public registerTemplate(templateId: string, template: string, options?: JSON): void {
    // TODO :: Add parials template support using options
    let preCompiledTemplate: Template[] = [];
    try {
      preCompiledTemplate = this.engine.parse(template);
    } catch (error) {
      throw new TemplateParseError(error.message);
    }
    this.preCompiledTemplateMap.set(templateId, preCompiledTemplate);
  }

  /**
   * Apply the template using the data provided.
   *
   * @param {string} templateId - id with which the compiled template is stored
   * @param {JSON} dataModel - data to apply to the template
   */
  public applyTemplate(templateId: string, dataModel: JSON): string {
    const preCompiledTemplate = this.preCompiledTemplateMap.get(templateId);
    if (!preCompiledTemplate) {
      throw new TemplateNotFound(`No template found for the given templateId : ${templateId}`);
    }
    return this.engine.renderSync(preCompiledTemplate, dataModel);
  }

  /**
   * Register custom helper functions with template engine.
   *
   * @param helperName name of the helper to register
   * @param helperFunc the implementation of helper function
   */
  // eslint-disable-next-line class-methods-use-this
  public registerHelper(helperName: string, helperFunc: FilterImplOptions): void {
    try {
      this.engine.registerFilter(helperName, helperFunc);
    } catch (error) {
      throw new CustomHelperRegisterError(`Registration of custom helper: ${helperName} failed with ERROR: ${error.message} `);
    }
  }

  /**
   * Register custom tag with template engine.
   *
   * @param tagName name of the tag to register
   * @param tagOptions tagOptions specific to the template engine
   */
  // eslint-disable-next-line class-methods-use-this
  public registerTag(tagName: string, tagOptions: TagImplOptions): void {
    try {
      this.engine.registerTag(tagName, tagOptions);
    } catch (error) {
      throw new CustomTagRegisterError(`Registration of custom Tag: ${tagName} failed with ERROR: ${error.message}`);
    }
  }

  private engine: Liquid;
  private preCompiledTemplateMap: Map<string, Template[]>;
}
