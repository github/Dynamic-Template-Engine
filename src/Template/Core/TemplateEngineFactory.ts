// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/event-transformer/blob/feature/chatops/LICENSE) for details)
import HandleBarsTemplateEngine from '../Engine/HandleBarsTemplateEngine';
import ITemplateEngine from './ITemplateEngine';
import { TemplateType } from '../../Transformer/Core/TransformContract';
import LiquidTemplateEngine from '../Engine/LiquidTemplateEngine';
import { TemplateEngineNotFound } from '../../Error/TemplateError';

/**
 * A singleton class
 * TemplateEngineFactory provides a method to get right engine based on template type
 */
export default class TemplateEngineFactory {
  private templateEngineMap: Map<TemplateType,
  ITemplateEngine> = new Map<TemplateType, ITemplateEngine>();
  private static instance: TemplateEngineFactory = new TemplateEngineFactory();

  /**
   * Private constructor sets up the templateEngineMap
   */
  private constructor() {
    this.templateEngineMap.set(TemplateType.HandleBars, new HandleBarsTemplateEngine());
    this.templateEngineMap.set(TemplateType.Liquid, new LiquidTemplateEngine());
  }

  /**
   * get instance of the singleton class
   *
   * @returns {TemplateEngineFactory} Instance of TemplateEngineFactory
   */
  public static getInstance(): TemplateEngineFactory {
    return this.instance;
  }

  /**
   * Gets template engine of the provided type
   *
   * @param {TemplateType} templateType - Type of the template engine, ex. Handlebars, Liquid
   * @returns {ITemplateEngine} template engine instance registered with the factory
   */
  public getTemplateEngine(templateType: TemplateType): ITemplateEngine {
    const templateEngine = this.templateEngineMap.get(templateType);
    if (!templateEngine) {
      throw new TemplateEngineNotFound(`No template engine present for the TemplateType: ${templateType}`);
    }
    return templateEngine;
  }
}
