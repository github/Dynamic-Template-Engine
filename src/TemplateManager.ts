// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/dynamic-template-engine/blob/master/LICENSE) for details)
import * as path from 'path';
import CustomEngineOptions from './Transformer/Model/CustomEngineOptions';
import TemplateEngineFactory from './Template/Core/TemplateEngineFactory';
import Transformer from './Transformer/Core/Transformer';
import TransformerConfig from './Transformer/Model/TransformerConfig';
import BaseTransformConfigEntry from './Transformer/Model/BaseTransformConfigEntry';
import CardRenderer from './Transformer/CardRenderer/CardRenderer';
import EventTransformer from './Transformer/EventTransformer/EventTransformer';
import { TemplateParseError, TemplateEngineNotFound } from './Error/TemplateError';
import { FileParseError, EmptyFileError } from './Error/FileError';
import CustomTemplatingOptions from './Transformer/Model/CustomTemplatingOptions';
import Utils from './Utility/Utility';
import { FunctionalityNotSupportedError, CustomHelperRegisterError, CustomTagRegisterError } from './Error/FunctionalityError';

/**
 * Template Manager provides methods to setup the template configuration
 * intializes template engines, registers all the templates provided by the config
 */
export default class TemplateManager {
  /**
   * Sets up the templates, picking from the path provided,
   * registering them to appropriate engines
   *
   * @param {string} configFilePath - path of the config file
   * @returns {boolean} true if setup successful
   * @throws Error if setup fails
   */
  public static async setupTemplateConfiguration(configFilePath: string,
    customOptions?: CustomTemplatingOptions): Promise<boolean> {
    try {
      const baseUrl = configFilePath.substring(0, configFilePath.lastIndexOf(path.sep));
      const transformerConfig = await this.readConfigFile(configFilePath, false);
      customOptions?.engineOptions?.forEach(engineOption => {
        this.registerHelpersAndTags(engineOption);
      });
      await this.registerAllTemplates(baseUrl, new CardRenderer(), transformerConfig.cardRenderer);
      await this.registerAllTemplates(baseUrl, new EventTransformer(),
        transformerConfig.eventTransformer);
    } catch (error) {
      const defaultMessage = `Setup failed for the config file path: ${configFilePath} \n ERROR: ${error.message}`;
      throw TemplateManager.handleErrors(error, defaultMessage);
    }
    return true;
  }

  /**
   * Sets up the templates, picking from the path provided,
   * registering them to appropriate engines
   *
   * @param {string} repo - repo name ex. user/repo
   * @param {string} branch - ex. master
   * @param {string} configName config file name in the repo root folder
   * @returns {boolean} true if setup succesful
   * @throws Error if setup fails
   */
  public static async setupTemplateConfigurationFromRepo(repo: string, branch: string,
    configName: string, customOptions?: CustomTemplatingOptions): Promise<boolean> {
    const baseUrl = `https://raw.githubusercontent.com/${repo}/${branch}`;
    try {
      const transformerConfig = await this.readConfigFile(`${baseUrl}/${configName}.json`, true);
      customOptions?.engineOptions?.forEach(engineOption => {
        this.registerHelpersAndTags(engineOption);
      });
      await this.registerAllTemplates(baseUrl, new CardRenderer(),
        transformerConfig.cardRenderer);
      await this.registerAllTemplates(baseUrl, new EventTransformer(),
        transformerConfig.eventTransformer);
    } catch (error) {
      const defaultMessage = `Setup failed for repo: ${repo} and branch: ${branch}. \n ERROR: ${error.message}`;
      throw TemplateManager.handleErrors(error, defaultMessage);
    }
    return true;
  }

  /**
   * Read config file and deserialize the file appropriately
   *
   * @param {string} filePath - file path of the config
   * @param {boolean} fromRepo - specifies if file from repo or from local machine
   */
  private static async readConfigFile(filePath: string,
    fromRepo: boolean): Promise<TransformerConfig> {
    const data = await Utils.fetchFile(fromRepo, filePath);
    try {
      return <TransformerConfig>JSON.parse(data.toString());
    } catch (error) {
      throw new FileParseError(`Unable to parse config file from path 
      ${filePath}, original error message: ${error.message}`);
    }
  }
  /**
   * Register all templates provided in the transformerConfig
   *
   * @param {string} baseUrl - base url for the template files
   * @param {string} transformer - transformer whith which template should be registered
   * @param {BaseTransformConfigEntry} transformerConfigs - the template transformer configs
   */
  private static async registerAllTemplates(baseUrl: string, transformer: Transformer<any>,
    transformerConfigs: BaseTransformConfigEntry[]): Promise<void> {
    // eslint-disable-next-line no-restricted-syntax
    for (const element of transformerConfigs) {
      try {
        // eslint-disable-next-line no-await-in-loop
        await transformer.registerTemplate(baseUrl, element);
      } catch (error) {
        if (error instanceof TemplateParseError) {
          throw new TemplateParseError(`Failed to parse template with name: ${element.TemplateName} 
          for source type: ${element.SourceType} and template type: ${element.TemplateType} with original message: ${error.message}`);
        } else {
          throw error;
        }
      }
    }
  }

  /**
   * Registers custom helpers and tags for a specific template engine
   *
   * @param {CustomEngineOptions} engineOption template type and the list of
   * custom helpers and tag to register
   */
  private static registerHelpersAndTags(engineOption: CustomEngineOptions): void {
    const engine = TemplateEngineFactory.getInstance().getTemplateEngine(engineOption.templateType);
    const helpers = engineOption.customHelpers;
    if (helpers) {
      Object.keys(helpers).forEach(helperName => {
        engine.registerHelper(helperName, helpers[helperName]);
      });
    }
    const tags = engineOption.customTags;
    if (tags) {
      Object.keys(tags).forEach(tagName => {
        engine.registerTag(tagName, tags[tagName]);
      });
    }
  }

  /**
   * Determine the correct error type and respond accordingly.
   *
   * @param {Error} error error to be analyzed
   * @param {string} defaultMessage message to show when error is not a known type
   * @returns {Error} correct error to throw
   */
  private static handleErrors(error: Error, defaultMessage: string): Error {
    if (error instanceof TemplateEngineNotFound || error instanceof TemplateParseError
      || error instanceof FileParseError || error instanceof EmptyFileError
      || error instanceof FunctionalityNotSupportedError
      || error instanceof CustomHelperRegisterError
      || error instanceof CustomTagRegisterError) {
      // eslint-disable-next-line @typescript-eslint/no-throw-literal
      return error;
    } else {
      return new Error(defaultMessage);
    }
  }
}
