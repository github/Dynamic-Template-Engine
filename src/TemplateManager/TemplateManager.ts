import * as fs from 'fs';
import * as https from 'https';
import TransformerConfig from "../TransformerConfig/TransformerConfig";
import { CardRenderer } from "../Transformer/CardRenderer/CardRenderer";
import TemplateEngineFactory from "../Templating/TemplateEngineFactory";
import CardRendererConfigEntry from "../Transformer/CardRenderer/CardRendererConfigEntry";
import EventTransformConfigEntry from "../Transformer/EventTransformer/EventTransformConfigEntry";
import PartialTemplateConfigEntry from "../Transformer/PartialTemplateConfigEntry";
import TemplateRegistrationHelper from './TemplateRegisterartionHelper';
import Utils from '../Utils/Utils';

// Template Manager provides methods to setup the template configuration
// intializes template engines, registers all the templates provided by the config
export default class TemplateManager{
  // Sets up the templates, picking from the path provided,
  // registering them to appropriate engines
  // Also initializes the template engines
  public static setupTemplateConfiguration(configFilePath: string){
    TemplateEngineFactory.initialize();
    try {
      const transformerConfig = this.readConfigFile(configFilePath);
      this.registerTemplates("", transformerConfig);
    } catch (error) {
      console.log(error);
    }
    console.log("Setup complete");
  }

  // Read config file and deserialize the file appropriately
  private static readConfigFile(filePath: string): TransformerConfig{
    const data = fs.readFileSync(filePath);
    return <TransformerConfig> JSON.parse(data.toString());
  }

  // Sets up the templates, picking from the repo provided,
  // registering them to appropriate engines
  // Also initializes the template engines
  public static async setupTemplateConfigurationFromRepo(repo: string, branch: string, configName: string){
    TemplateEngineFactory.initialize();
    const baseUrl = `https://raw.githubusercontent.com/${repo}/${branch}`;
    try {
      const transformerConfig = await this.readConfigFileFromRepo(`${baseUrl}/${configName}.json`);
      this.registerTemplates(baseUrl, transformerConfig);
    } catch(error) {
      console.log(error);
    }
    
  }

    // Read config file from repo and deserialize the file appropriately
  private static async readConfigFileFromRepo(filePath: string): Promise<TransformerConfig>{
    const data = await Utils.httpSync(filePath);
    console.log(`data from repo : ${data}`);
    return <TransformerConfig> JSON.parse(data);
  }

  // Register all templates provided in the transformerConfig  
  private static registerTemplates(baseUrl:string, transformerConfig: TransformerConfig) {
    const isHttpCall = baseUrl.length > 0;
    const basepath = isHttpCall ? `${baseUrl}/CardTemplates` : 'CardTemplates';
    if(transformerConfig){
      this.registerCardRenderers(basepath, transformerConfig.cardRenderers, isHttpCall);
      this.registerEventTransformers(basepath, transformerConfig.eventTransformers, isHttpCall);
      this.registerPartials(basepath, transformerConfig.partials, isHttpCall);
    }
  }

  // TODO :: Implement Partials support 
  private static registerPartials(basepath: string, partials: PartialTemplateConfigEntry[], isHttpCall: boolean) {
    if(partials){
      partials.forEach(()=>{
        // TODO :: code for registering partials with template engine
      });
    }
  }

  // TODO :: Implement Event Transformers 
  private static registerEventTransformers(basepath: string, eventTransformers: EventTransformConfigEntry[], isHttpCall: boolean) {
    if(eventTransformers){
      eventTransformers.forEach(()=>{
        // TODO :: code for registering event transformers 
      });
    }
  }

  // Registers card renders picking the templates from the basePath and and card renderers from the Configs passed
  private static registerCardRenderers(basepath: string, cardRenderers: CardRendererConfigEntry[], isHttpCall: boolean){
    if(cardRenderers){
      cardRenderers.forEach(async element => {
        try {
          const path = `${basepath}/${element.ClientType}/${element.TemplateType}/${element.TemplateName}`;
          const templateFile = isHttpCall? await Utils.httpSync(path) : fs.readFileSync(path).toString();
          TemplateRegistrationHelper.registerCardRenderer(element.TemplateType, element.SourceType,element.ClientType, templateFile);
        } catch (error) {
          console.log(error); 
        }
      });
    }
  }
}