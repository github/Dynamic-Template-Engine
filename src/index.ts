// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/dynamic-template-engine/blob/master/LICENSE) for details)
import TemplateManager from './TemplateManager';
import CardRenderer from './Transformer/CardRenderer/CardRenderer';
import EventTransformer from './Transformer/EventTransformer/EventTransformer';
import CustomEngineOptions from './Transformer/Model/CustomEngineOptions';
import CustomTemplatingOptions from './Transformer/Model/CustomTemplatingOptions';

export { TemplateManager };
export { CardRenderer, EventTransformer };
export { CustomEngineOptions, CustomTemplatingOptions };
export { ClientType, TemplateType } from './Transformer/Core/TransformContract';
