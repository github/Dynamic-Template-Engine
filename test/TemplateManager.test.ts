import TemplateManager from '../src/TemplateManager';
import Utility from '../src/Utility/Utility';
import CardRenderer from '../src/Transformer/CardRenderer/CardRenderer';
import EventTransformer from '../src/Transformer/EventTransformer/EventTransformer';
import { FileReadError } from '../src/Error/FileError';
import { TemplateParseError } from '../src/Error/TemplateError';
import HandleBarsTemplateEngine from '../src/Template/Engine/HandleBarsTemplateEngine';
import LiquidTemplateEngine from '../src/Template/Engine/LiquidTemplateEngine';
import { TemplateType, CustomEngineOptions, CustomTemplatingOptions } from '../src';
import TemplateEngineFactory from '../src/Template/Core/TemplateEngineFactory';

jest.mock('../src/Transformer/CardRenderer/CardRenderer');
jest.mock('../src/Transformer/EventTransformer/EventTransformer');
jest.mock('../src/Utility/Utility');
jest.mock('../src/Template/Engine/HandleBarsTemplateEngine');
jest.mock('../src/Template/Engine/LiquidTemplateEngine');

const CardRendererMock: jest.Mock<CardRenderer> = <jest.Mock<CardRenderer>>(CardRenderer as any);
const EventTransformerMock: jest.Mock<EventTransformer> = <jest.Mock<EventTransformer>>(EventTransformer as any);
const HandleBarsTemplateEngineMock: jest.Mock<HandleBarsTemplateEngine> = <jest.Mock<HandleBarsTemplateEngine>>(HandleBarsTemplateEngine as any);
const LiquidTemplateEngineMock: jest.Mock<LiquidTemplateEngine> = <jest.Mock<LiquidTemplateEngine>>(LiquidTemplateEngine as any);

const mockTransformerConfig = {
  "cardRenderer": [
    {
      "SourceType": "IssueOpened",
      "ClientType": "Teams",
      "TemplateType": "HandleBars",
      "TemplateName": "issue_opened.handlebars"
    }
  ],
  "partials": [],
  "eventTransformer": [
    {
      "SourceType": "IssueOpened",
      "TemplateType": "HandleBars",
      "TemplateName": "issue_opened.handlebars"
    }
  ]
}

const handlebarsEngineOptions: CustomEngineOptions = {
  templateType: TemplateType.HandleBars,
  customHelpers: {
    'upperCaseTest': (str: string) => { return str.toUpperCase() },
    'lowerCaseTest': (str: string) => { return str.toLowerCase() }
  }
};

const liquidEngineOptions: CustomEngineOptions = {
  templateType: TemplateType.Liquid,
  customHelpers: {
    'upperCaseTest': (str: string) => { return str.toUpperCase() },
    'lowerCaseTest': (str: string) => { return str.toLowerCase() }
  },
  customTags: {
    'upperTest': {
      parse: function(this: any,tagToken: any, remainTokens: any): void {
        this.str1 = tagToken.args;
      },
      render: function(this:any, scope: any, hash: any): string {
        let str = scope.environments[this.str1];
        return str.toUpperCase(); 
      }
    }
  }
};

const onlyTagsLiquidOptions: CustomEngineOptions = {
  templateType: TemplateType.Liquid,
  customTags: liquidEngineOptions.customTags
}

const onlyHelpersLiquidOptions: CustomEngineOptions = {
  templateType: TemplateType.Liquid,
  customHelpers: liquidEngineOptions.customHelpers
}

describe('TemplateManager Unit Tests', () => {

  beforeAll(() => {
    jest.spyOn(TemplateManager as any, 'registerAllTemplates');
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('setupTemplateConfiguration fails for no file', () => {
    expect(TemplateManager.setupTemplateConfiguration('nonExistantFilePath')).rejects.toThrowErrorMatchingSnapshot();
  });

  it('setupTemplateConfigurationFromRepo fails for no file', () => {
    expect(TemplateManager.setupTemplateConfigurationFromRepo('nonExistantrepo', 'nonExistantBranch', 'nonExistantFileName')).rejects.toThrowErrorMatchingSnapshot();
  });

  it('setupTemplateConfiguration registers all templates', async () => {
    Utility.fetchFile = jest.fn().mockReturnValueOnce(JSON.stringify(mockTransformerConfig));
    await TemplateManager.setupTemplateConfiguration('mockTransformerConfig');
    expect((TemplateManager as any).registerAllTemplates).toHaveBeenCalledTimes(2);
    expect(CardRendererMock.mock.instances[0].registerTemplate).toHaveBeenCalledTimes(1);
    expect(EventTransformerMock.mock.instances[0].registerTemplate).toHaveBeenCalledTimes(1);
  });

  it('setupTemplateConfigurationFromRepo registers all templates', async () => {
    Utility.fetchFile = jest.fn().mockReturnValueOnce(JSON.stringify(mockTransformerConfig));
    await TemplateManager.setupTemplateConfigurationFromRepo('mockRepo', 'mockBranch', 'mockFile');
    expect((TemplateManager as any).registerAllTemplates).toHaveBeenCalledTimes(2);
    expect(CardRendererMock.mock.instances[0].registerTemplate).toHaveBeenCalledTimes(1);
    expect(EventTransformerMock.mock.instances[0].registerTemplate).toHaveBeenCalledTimes(1);
  });

  // to know more about usage of each https://jestjs.io/docs/en/api#testeachtablename-fn-timeout
  it.each`ErrorToThrow | ErrorExpected
  ${TemplateParseError} | ${TemplateParseError}
  ${FileReadError} | ${Error}`('setupTemplateConfigurationFromRepo fails if any register template fails', async ({ ErrorToThrow, ErrorExpected }) => {
    Utility.fetchFile = jest.fn()
      .mockImplementationOnce(async (isHttpCall, filePath) => { return JSON.stringify(mockTransformerConfig) });
    jest.spyOn(CardRenderer.prototype, 'registerTemplate').mockImplementationOnce(() => {
      throw new ErrorToThrow('Test');
    });
    try {
      await TemplateManager.setupTemplateConfigurationFromRepo('mockRepo', 'mockBranch', 'mockFile');
    } catch (e) {
      expect(e).toBeInstanceOf(ErrorExpected);
    }

  });

  it('original message should not be lost if setup process fails', async () => {
    const originalMessage = 'Original message of the error';
    jest.spyOn(TemplateManager as any, 'readConfigFile').mockImplementationOnce(() => { throw new Error(originalMessage) });
    try {
      await TemplateManager.setupTemplateConfiguration('mockTransformerConfig');
    } catch (e) {
      expect(e.message).toContain(originalMessage);
    }
  });

  it('original message should not be lost if setup process fails', async () => {
    const originalMessage = 'Original message of the error';
    jest.spyOn(TemplateManager as any, 'readConfigFile').mockImplementationOnce(() => { throw new Error(originalMessage) });
    try {
      await TemplateManager.setupTemplateConfigurationFromRepo('mockRepo', 'mockBranch', 'mockConfigFile');
    } catch (e) {
      expect(e.message).toContain(originalMessage);
    }
  });

  // to know more about usage of each https://jestjs.io/docs/en/api#testeachtablename-fn-timeout
  it.each`
  testType | options 
  ${'test register Helpers for handlebars'} | ${handlebarsEngineOptions}
  ${'test register tags and helpers for liquid'} | ${liquidEngineOptions}
  ${'test register only helpers for liquid'} | ${onlyHelpersLiquidOptions}
  ${'test register only tags for liquid'} | ${onlyTagsLiquidOptions}
  `('registerHelpersAndTags: $testType', ({testType, options}) => {
    (TemplateManager as any).registerHelpersAndTags(options);
    let engine = TemplateEngineFactory.getInstance().getTemplateEngine(options.templateType);
    if (options.customHelpers){
      let size = Object.keys(options.customHelpers).length;
      expect(engine.registerHelper).toHaveBeenCalledTimes(size);
    }
    if (options.customTags){
      let size = Object.keys(options.customTags).length;
      expect(engine.registerTag).toHaveBeenCalledTimes(size);
    }
  });

});