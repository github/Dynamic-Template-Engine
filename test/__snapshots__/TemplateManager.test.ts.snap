// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateManager Unit Tests setupTemplateConfiguration fails for no file 1`] = `
"Unable to parse config file from path 
      nonExistantFilePath, original error message: Cannot read property 'toString' of undefined"
`;

exports[`TemplateManager Unit Tests setupTemplateConfigurationFromRepo fails for no file 1`] = `
"Unable to parse config file from path 
      https://raw.githubusercontent.com/nonExistantrepo/nonExistantBranch/nonExistantFileName.json, original error message: Cannot read property 'toString' of undefined"
`;
